lib/jsdom/living/xhr/XMLHttpRequest-impl.js

```diff
@@ -0,0 +1,1109 @@
+const { isArrayBuffer } = require("../generated/utils");
+const { parseIntoDocument } = require("../../browser/parser");
+const { fragmentSerialization } = require("../domparsing/serialization");
+const { setupForSimpleEventAccessors } = require("../helpers/create-event-accessor");
+const { parseJSONFromBytes } = require("../helpers/json");
+const { fireAnEvent } = require("../helpers/events");
+const { copyToArrayBufferInNewRealm } = require("../helpers/binary-data");
+
+/**
+ * @typedef {import("../file-api/Blob-impl.js").implementation} BlobImpl
+ * @typedef {import("./FormData-impl.js").implementation} FormDataImpl
+ * @typedef {import("../nodes/Document-impl.js")["implementation"]["prototype"]} DocumentImpl
+ *
+ * @typedef {string | BlobImpl | BufferSource | FormDataImpl } BodyInitImpl
+ * @typedef {NodeJS.EventEmitter & {
+  abort(): void;
+  statusCode?: number;
+  rawHeaders?: string[];
+  headers?: { [header: string]: string };
+  request?: { uri: import("whatwg-url").URL };
+  response?: import("request").Response;
+}} Client
+ */
+
+const syncWorkerFile = require.resolve ? require.resolve("./xhr-sync-worker.js") : null;
+
+const tokenRegexp = /^[!#$%&'*+\-.^_`|~0-9A-Za-z]+$/;
+const fieldValueRegexp = /^[ \t]*(?:[\x21-\x7E\x80-\xFF](?:[ \t][\x21-\x7E\x80-\xFF])?)*[ \t]*$/;
+
+const forbiddenRequestHeaders = new Set([
+  "accept-charset",
+  "accept-encoding",
+  "access-control-request-headers",
+  "access-control-request-method",
+  "connection",
+  "content-length",
+  "cookie",
+  "cookie2",
+  "date",
+  "dnt",
+  "expect",
+  "host",
+  "keep-alive",
+  "origin",
+  "referer",
+  "te",
+  "trailer",
+  "transfer-encoding",
+  "upgrade",
+  "via"
+]);
+const forbiddenResponseHeaders = new Set([
+  "set-cookie",
+  "set-cookie2"
+]);
+const uniqueResponseHeaders = new Set([
+  "content-type",
+  "content-length",
+  "user-agent",
+  "referer",
+  "host",
+  "authorization",
+  "proxy-authorization",
+  "if-modified-since",
+  "if-unmodified-since",
+  "from",
+  "location",
+  "max-forwards"
+]);
+const corsSafeResponseHeaders = new Set([
+  "cache-control",
+  "content-language",
+  "content-type",
+  "expires",
+  "last-modified",
+  "pragma"
+]);
+
+const allowedRequestMethods = new Set(["OPTIONS", "GET", "HEAD", "POST", "PUT", "DELETE"]);
+const forbiddenRequestMethods = new Set(["TRACK", "TRACE", "CONNECT"]);
+
+/** @type {ReadonlySet<XMLHttpRequestResponseType>} */
+const XMLHttpRequestResponseType = new Set([
+  "",
+  "arraybuffer",
+  "blob",
+  "document",
+  "json",
+  "text"
+]);
+
+class XMLHttpRequest extends XMLHttpRequestEventTargetImpl {
+  constructor(window) {
+    super(window);
+
+    /** @type {import("../../browser/Window")} */
+    this._globalObject = window;
+    /** @type {import("./XMLHttpRequestUpload-impl.js").implementation} */
+    this.upload = XMLHttpRequestUpload.createImpl(window);
+    this.upload._ownerDocument = this._ownerDocument;
```

Comment:
I don't see this being used anywhere that's obvious… is this necessary? In either case it seems to belong more in `XMLHttpRequestUpload`'s impl constructor.